lightning-net/src/lib.rs

```diff
@@ -0,0 +1,1098 @@
+    peer_manager: Arc<PeerManager<SyncSocketDescriptor, Arc<CMH>, Arc<RMH>, Arc<L>, Arc<UMH>>>,
+    descriptor: SyncSocketDescriptor,
+    reader_cmd_rx: Receiver<ReaderCommand>,
+    state: ReaderState,
+}
+impl<CMH, RMH, L, UMH> Reader<CMH, RMH, L, UMH>
+where
+    CMH: ChannelMessageHandler + 'static + Send + Sync,
+    RMH: RoutingMessageHandler + 'static + Send + Sync,
+    L: Logger + 'static + ?Sized + Send + Sync,
+    UMH: CustomMessageHandler + 'static + Send + Sync,
+{
+    fn new(
+        reader: TcpReader,
+        peer_manager: Arc<PeerManager<SyncSocketDescriptor, Arc<CMH>, Arc<RMH>, Arc<L>, Arc<UMH>>>,
+        descriptor: SyncSocketDescriptor,
+        reader_cmd_rx: Receiver<ReaderCommand>,
+    ) -> Self {
+        Self {
+            inner: reader,
+            peer_manager,
+            descriptor,
+            reader_cmd_rx,
+            state: ReaderState::Reading,
+        }
+    }
+
+    /// Handle read events, or wait for the next `ReaderCommand` if reads are
+    /// paused. This implementation avoids busy loops and lets the thread go to
+    /// sleep whenever reads or channel commands are pending.
+    fn run(&mut self) {
+        use ReaderState::*;
+
+        // 8KB is nice and big but also should never cause any issues with stack
+        // overflowing.
+        let mut buf = [0; 8192];
+
+        loop {
+            self.do_try_recv();
+
+            match self.state {
+                Reading => self.do_read(&mut buf),
+                Paused => self.do_recv(),
+                ShuttingDown => break,
+            };
+        }
+
+        // Shut down the underlying stream. It's fine if it was already closed.
+        let _ = self.inner.shutdown();
+        // Send a signal to the Writer to do the same.
+        let _ = self
+            .descriptor
+            .writer_cmd_tx
+            .try_send(WriterCommand::Shutdown);
+    }
+
+    /// Checks for a command in a non-blocking manner, handling the command
+    /// if there was one.
+    fn do_try_recv(&mut self) {
+        match self.reader_cmd_rx.try_recv() {
+            Ok(cmd) => self.handle_command(cmd),
+            Err(e) => match e {
+                TryRecvError::Empty => {}
+                TryRecvError::Disconnected => self.state = ReaderState::ShuttingDown,
+            },
+        };
+    }
+
+    /// Blocks on the command channel and handles the command.
+    fn do_recv(&mut self) {
+        match self.reader_cmd_rx.recv() {
+            Ok(cmd) => self.handle_command(cmd),
+            Err(_) => self.state = ReaderState::ShuttingDown,
+        };
+    }
+
+    /// Handles a `ReaderCommand`.
+    fn handle_command(&mut self, cmd: ReaderCommand) {
+        use ReaderCommand::*;
+        use ReaderState::*;
+
+        match (cmd, &self.state) {
+            (_, &ShuttingDown) => {}
+            (PauseRead, _) => self.state = Paused,
+            (ResumeRead, _) => self.state = Reading,
+            (Shutdown, _) => self.state = ShuttingDown,
```

Comment:
The `Shutdown` message was implemented because while reading is paused, we don't want the `Reader` to be, well, reading. The reader needs to block on _something_, otherwise there is no way for it to know to resume reading (or to shut down, for that matter) without a busy loop. So that's why the `reader_cmd` channel is necessary. But since the Reader can be blocked on the channel instead of `read()`, it would miss any `TcpStream::shutdown()`s initiated by the `Writer` or `PeerManager`. Thus, the `Shutdown` command which can unblock the `Reader` while it is stuck on `recv()`.

Reply:
The doc comment for `SyncSocketDescriptor::disconnect_socket()` explains why the current design uses channels:

(The below is after fixing a typo)
```
    /// Shuts down the Reader, Writer, and the underlying TcpStream.
    ///
    /// There are several ways that a disconnect might be triggered:
    /// 1) The Reader receives Ok(0) from TcpStream::read() (i.e. the
    ///    peer disconnected), or an Err(io::Error) that shouldn't be retried.
    /// 2) The Reader receives Err from PeerManager::read_event(); i.e.
    ///    Rust-Lightning told us to disconnect from the peer.
    /// 3) The Writer receives Ok(0) from TcpStream::write() (undocumented
    ///    behavior), or an Err(io::Error) that shouldn't be retried.
    /// 4) The Writer receives Err from PeerManager::write_buffer_space_avail();
    ///    Rust-Lightning told us to disconnect from the peer.
    /// 5) This function is called.
    ///
    /// The disconnect will be handled differently depending on the source of
    /// the trigger:
    /// - (1) and (2): If the Reader received the trigger, it will shut down
    ///   BOTH halves of the shared TcpStream AND send a Shutdown command to the
    ///   Writer.
    ///
    ///   - The explicit Shutdown command from the Reader is necessary because
    ///     if the Writer is blocked on `writer_cmd_rx.recv()` due to its
    ///     internal buffer being empty, the only way it can be unblocked is by
    ///     receiving a command, in this case the Shutdown command.
    ///   - The Reader closing both halves of the TCP stream is necessary
    ///     because while the writer is blocked on write(), the only way it can
    ///     unblock is by detecting the TCP disconnect.
    ///
    /// - (3) and (4): If the Writer received the trigger, it will shut down
    ///   BOTH halves of the shared TcpStream AND send a Shutdown command to the
    ///   Reader.
    ///
    ///   - The explicit Shutdown command from the Writer is necessary because
    ///     if the Reader is blocked on `reader_cmd_rx.recv()` due to
    ///     `read_paused == true`, the only way it can be unblocked is by
    ///     receiving a command, in this case the Shutdown command.
    ///   - The Writer closing both halves of the TCP stream is necessary
    ///     because while the reader is blocked on read(), the only way it can
    ///     unblock is by detecting the TCP disconnect.
    ///
    /// - (5): If the disconnect was initiated here, a Shutdown command will be
    ///   sent to both the Reader and the Writer, AND the TcpDisconnectooor will
    ///   shutdown both halves of the shared TCP stream. The Shutdown command
    ///   ensures that the Reader / Writer will unblock if they are currently
    ///   blocked on `recv()`. The TCP stream shutdown ensures that they will
    ///   unblock if they are currently blocked on `read()` or `write()`
    ///   respectively.
    ///
    /// In cases (1) and (3), the disconnect was NOT initiated by
    /// Rust-Lightning, so the Reader / Writer notify the PeerManager using
    /// `socket_disconnected()`.
```

Reply:
Re `TcpStream::shutdown()` noop in the SGX docs - nice attention to detail, you actually read the docs! Indeed, the code in its current state will not function correctly inside SGX. The `Socket` trait that we plan to introduce in the will will have a `shutdown()` method. That way, we can swap out the `std::net::TcpStream` with our own `SgxTcpStream` which uses an EDP usercall to implement `shutdown()`.

The remaining comments about `std::time` and `std::thread` are in the `lightning-net` doc comments so that future implementers know to avoid using those in the rest of the crate, since the `SgxTcpStream` will still rely on this code.